File: RFID Reader/手持终端1-AUTOID9U（东集）/开发文档_V4.1.0.R5/Source/VBFormDemo/MainForm.vb

Imports System.IO
Imports System.Security.Cryptography
Imports System.Runtime.InteropServices
Imports System.Collections
Imports System.Collections.Generic
Imports System.IO.Ports
Imports System.Threading
Imports JW.UHF

Public Class MainForm
    Dim TimeControlThread As System.Threading.Thread
    Public Shared StopInventory As Boolean = False
    Delegate Sub ControlHandler(ByVal message As String)
    Dim tagList As New Dictionary(Of String, ListViewItem)

    Dim jwReader As JWReader
    Private inventoryThread As Thread
    Private updateUIThread As Thread
    Private lockObj As New Object()
    Dim tagQueue As New Queue()

    Private ReadCount As Integer = 0, TagCount As Integer = 0, TagAmount As Integer = 0, inventoryTime As Integer = 0

    Public Sub displog(ByVal msg As String)
        ListBox_Log.Items.Add(msg)
    End Sub

    Friend Delegate Sub ShowListViewCallBack(Tag As Tag)

    Friend Sub UpdateView(Tag As Tag)

        Dim sEPC As String = Tag.EPC
        Dim sIP As String = ""
        Dim sRSSI As Single = Tag.RSSI
        Dim sAnt As Integer = Tag.PORT
        ' 年月日時分秒毫秒
        Dim sTime As String = DateTime.Now.ToString("yyyy/MM/dd hh:mm:ss:ffff")

        If tagList.ContainsKey(sEPC) Then
            Dim listRow As ListViewItem = tagList.Item(sEPC)
            Dim updateRow As ListViewItem
            For Each row As ListViewItem In ListViewInfo.Items
                If listRow Is row Then
                    updateRow = row
                    Exit For
                End If
            Next

            If Not updateRow Is Nothing Then

                Dim s As String = updateRow.SubItems.Item(1).Text
                Dim countValue As UInt64 = UInt64.Parse(s) + 1
                updateRow.SubItems.Item(1).Text = countValue.ToString()
                updateRow.SubItems.Item(2).Text = sAnt.ToString()
                updateRow.SubItems.Item(3).Text = sRSSI.ToString()
                updateRow.SubItems.Item(4).Text = sAnt.ToString()
                updateRow.SubItems.Item(5).Text = sTime
            End If

        Else

            TagAmount = ListViewInfo.Items.Count

            Dim item As ListViewItem = New ListViewItem(sEPC)
            item.SubItems.Add("1")
            item.SubItems.Add(sAnt.ToString())
            item.SubItems.Add(sRSSI.ToString())
            item.SubItems.Add(sAnt.ToString())
            item.SubItems.Add(sTime)
            ListViewInfo.Items.Add(item)
            tagList.Add(sEPC, item)
            'this.actual_read_count++;
            'this.TagCountTb.Text = this.actual_read_count.ToString();
            'this.TagCountTb.Update();

        End If
    End Sub
    Friend Sub ShowListView()
        While Not StopInventory

            Dim WriteFlag As Boolean = True
            Dim dt As DateTime = DateTime.Now
            Dim ReadTime As Int64
            ReadTime = Int64.Parse(String.Format("{0:yyyyMMddHHmmssffff}", dt))


            While tagQueue.Count > 0
                Dim tag As Tag

                SyncLock lockObj
                    tag = tagQueue.Dequeue
                End SyncLock

                Me.Invoke(New ShowListViewCallBack(AddressOf UpdateView), tag)

            End While
        End While

    End Sub

    Public Sub ClearLog()
        ListBox_Log.Items.Clear()
        ListViewInfo.Items.Clear()
        tagList.Clear()
        tagQueue.Clear()
        TextBox1.Text = ""
    End Sub

    Public Function GetAvaliableSerialPorts() As String()
        Dim portNames() As String = SerialPort.GetPortNames()
        Array.Sort(portNames)
        Return portNames
    End Function


    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Dim portNames() As String = GetAvaliableSerialPorts()

        For Each s In portNames
            ComboBox1.Items.Add(s)
        Next

        If ComboBox1.Items.Count > 0 Then
            ComboBox1.SelectedIndex = 0
        End If

    End Sub

    Public Sub TagsReport(sender As JWReader, args As TagsEventArgs)
        Dim tag As Tag = args.tag
        SyncLock lockObj
            tagQueue.Enqueue(tag)
        End SyncLock

        'Console.WriteLine(String.Format("EPC={0},Port={1},RSSI={2}", tag.EPC, tag.PORT, tag.RSSI))
        'ShowListView(tag.EPC, "", tag.RSSI, tag.PORT, Now.ToShortTimeString)

    End Sub



    Private Sub Btn_Start_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Btn_Start.Click

        StopInventory = False
        tagQueue.Clear()

        displog("Start ..." + ComboBox1.SelectedItem.ToString)

        Application.DoEvents()

        Dim key As Integer = 0
        jwReader = New JWReader(ComboBox1.Text)
        'jwReader = New JWReader("10.10.10.64", 9761)
        Dim result As Result = jwReader.RFID_Open()

        If result <> result.OK Then
            displog("Open RFID module fail, please try other config..")
            Exit Sub
        End If
        displog("Open UHF Module Successful ...")

        Dim rs As RfidSetting = New RfidSetting()
        rs.AntennaPort_List = New List(Of AntennaPort)()
        Dim ap1 As AntennaPort = New AntennaPort()
        Dim ap2 As AntennaPort = New AntennaPort()
        Dim ap3 As AntennaPort = New AntennaPort()
        Dim ap4 As AntennaPort = New AntennaPort()

        'If ap1.Exist = True Then
        If CheckBox1.Checked = True Then
            ap1.AntennaIndex = 0 '天线0
            ap1.Power = 22       '功率
            rs.AntennaPort_List.Add(ap1)

            displog("天線1設定完成!")
        End If
        'End If

        'If ap2.Exist = True Then
        If CheckBox2.Checked = True Then
            ap2.AntennaIndex = 1 '天线1
            ap2.Power = 22       '功率
            rs.AntennaPort_List.Add(ap2)

            displog("天線2設定完成!")
        End If
        'End If

        'If ap3.Exist = True Then
        If CheckBox3.Checked = True Then
            ap3.AntennaIndex = 2 '天线2
            ap3.Power = 22       '功率
            rs.AntennaPort_List.Add(ap3)

            displog("天線3設定完成!")
            'End If
        End If


        'If ap4.Exist = True Then
        If CheckBox4.Checked = True Then
            ap4.AntennaIndex = 3 '天线3
            ap4.Power = 22       '功率
            rs.AntennaPort_List.Add(ap4)

            displog("天線4設定完成!")
        End If
        'End If

        rs.GPIO_Config = Nothing
        displog("Get module configuration successfully")

        rs.Inventory_Time = 0

        rs.Region_List = RegionList.CCC

        'rs.RSSI_Filter = New RSSIFilter()
        'rs.RSSI_Filter.Enable = True
        'rs.RSSI_Filter.RSSIValue = -70

        rs.Speed_Mode = SpeedMode.SPEED_FASTEST


        rs.Tag_Group = New TagGroup()
        rs.Tag_Group.SessionTarget = SessionTarget.A
        rs.Tag_Group.SearchMode = SearchMode.DUAL_TARGET
        rs.Tag_Group.Session = Session.S0

        result = jwReader.RFID_Set_Config(rs)
        If result = result.OK Then
            displog("RFID Config Set Success")
        Else
            displog("RFID Config Set Failure")
            result = jwReader.RFID_Close()
            Return
        End If


        'jwReader.TagsReported += TagsReport()
        AddHandler jwReader.TagsReported, AddressOf Me.TagsReport

        inventoryThread = New Thread(AddressOf ThreadTask)
        inventoryThread.IsBackground = True
        inventoryThread.Start()

        updateUIThread = New Thread(AddressOf ShowListView)
        updateUIThread.IsBackground = True
        updateUIThread.Start()


        displog("Start Inventory")




    End Sub

    Private Sub ThreadTask()
        jwReader.RFID_Start_Inventory()
    End Sub

    Private Sub Btn_Cls_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Btn_Cls.Click
        ClearLog()
    End Sub

    Private Sub Btn_Stop_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Btn_Stop.Click

        StopInventory = True

        jwReader.RFID_Stop_Inventory()
        displog("Stop Inventory")

        jwReader.RFID_Close()
        displog("Close RFID Reader")

    End Sub

End Class

File: RFID Reader/东集autoid9u手持终端/Source/WinFormDemo/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Threading;
using JW.UHF;
namespace WinFormDemo
{
    public partial class MainForm : Form
    {

        private object lockObj = new object();//线程同步锁
        private DateTime startTime;//启动时间 
        private Queue<Tag> inventoryTagQueue = new Queue<Tag>();//盘点到Tag队列列表
        Dictionary<string, ListViewItem> tagList = new Dictionary<string, ListViewItem>();//Tag列表
        UInt64 actual_read_count = 0;//实际读取数量
        private bool stopInventoryFlag = false;//是否停止盘点标志


        private delegate void UHFOperDelegate();//UHF操作跨线程委托类

        public MainForm()
        {
            InitializeComponent();
        }

        private JWReader jwReader = null;
        /// <summary>
        /// 应用启动
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void MainForm_Load(object sender, EventArgs e)
        {
            CustomTraceListener.Show();
            #region 连接模块
            Result result = Result.OK;
            jwReader = new JWReader("COM4");
            //jwReader = new JWReader("192.168.100.168",9761);

            result = jwReader.RFID_Open();//连接UHF模块

            if (result != Result.OK)
            {
                #region 第二次尝试打开模块
                result = jwReader.RFID_Open();
                if (result != Result.OK)
                {
                    MessageBox.Show("Open UHF Module Failure");
                    this.Close();
                }
                #endregion

            }
            #endregion

            //result=jwReader.
            #region 配置模块
            RfidSetting rs = new RfidSetting();
            rs.AntennaPort_List = new List<AntennaPort>();
            
            AntennaPort ap = new AntennaPort();
            ap.AntennaIndex = 0;//天线0
            ap.Power = 22;//功率设为22
            rs.AntennaPort_List.Add(ap);
           

            rs.GPIO_Config = null;


            rs.Inventory_Time = 5000;

            rs.Region_List = RegionList.CCC;

            rs.RSSI_Filter = new RSSIFilter();
            rs.RSSI_Filter.Enable = true;
            rs.RSSI_Filter.RSSIValue = (float)-70;

            rs.Speed_Mode = SpeedMode.SPEED_FASTEST;


            rs.Tag_Group = new TagGroup();
            rs.Tag_Group.SessionTarget = SessionTarget.A;
            rs.Tag_Group.SearchMode = SearchMode.DUAL_TARGET;
            rs.Tag_Group.Session = Session.S0;

            result = jwReader.RFID_Set_Config(rs);
            if (result != Result.OK)
            {
                MessageBox.Show("RFID Config Set Failure");
                this.Close();
            }
       
            #endregion 
            this.stopInventoryBt.Enabled = false;
        }

        /// <summary>
        /// 应用关闭
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void MainForm_Closing(object sender, CancelEventArgs e)
        {
            Result result = Result.OK;

            result = jwReader.RFID_Stop_Inventory();//停止当前UHF操作

            result = jwReader.RFID_Close();//关闭模块连接

            
        }

        /// <summary>
        /// 清空盘点数据
        /// </summary>
        private void clearInventoryData()
        {
            inventoryTagQueue.Clear();
            tagList.Clear();
            this.tagListView.Items.Clear();
            actual_read_count = 0;
            this.TagCountTb.Text = "0";
            this.consumeTimeTb.Text = "";
        }


        /// <summary>
        /// 启动盘点
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void startBtn_Click(object sender, EventArgs e)
        {
            clearInventoryData();//清空盘点数据

            stopInventoryFlag = false;

            Thread inventoryThread = new Thread(inventory);//盘点线程
            inventoryThread.Start();

            Thread updateThread = new Thread(updateList);//更新列表线程
            updateThread.Start();
        }


        /// <summary>
        /// 数据上报
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="args"></param>
        private void TagsReport(object sender, TagsEventArgs args)
        {
            Tag tag = args.tag;
            lock (lockObj)
                inventoryTagQueue.Enqueue(tag);//回调函数事情越少越好。
        }


        /// <summary>
        /// 盘点线程
        /// </summary>
        void inventory()
        {
            //盘点开始初始化动作
            UHFOperDelegate modifyInventoryStart = delegate()
            {
                this.stopInventoryBt.Enabled = true;
                this.startInventoryBt.Enabled = false;
                this.clearBtn.Enabled = false;

                startTime = DateTime.Now;
            };
            this.startInventoryBt.Invoke(modifyInventoryStart);

            jwReader.TagsReported += TagsReport;
            //盘点
            jwReader.RFID_Start_Inventory();

            //盘点完毕处理动作
            DateTime endTime = DateTime.Now;
            UHFOperDelegate modifyInventoryStop = delegate()
            {
                this.startInventoryBt.Enabled = true;
                this.stopInventoryBt.Enabled = false;
                this.clearBtn.Enabled = true;
                this.stopInventoryFlag = true;
                double consumeTime = Util.DateDiffMillSecond(endTime, startTime);//毫秒
                this.consumeTimeTb.Text = (consumeTime / 1000).ToString("0");
            };
            this.stopInventoryBt.Invoke(modifyInventoryStop);

        }



        /// <summary>
        /// 更新列表线程
        /// </summary>
        /// <param name="tags"></param>
        private void updateList()
        {
            while (!stopInventoryFlag)//未停止
            {
                updateInventoryGridList();
                Thread.Sleep(100);
            }
            
            DateTime dt = DateTime.Now;
            while (true)
            {
                updateInventoryGridList();
                //500毫秒内确定没有包了 防止线程提前结束 有些盘点包还没处理完 可保证该线程最后结束。
                if (inventoryTagQueue.Count == 0 && Util.DateDiffMillSecond(DateTime.Now, dt) > 500)
                    break;
            }
            
        }

        /// <summary>
        /// 更新列表
        /// </summary>
        private void updateInventoryGridList()
        {
            UHFOperDelegate updateList = delegate()
            {
                while (inventoryTagQueue.Count > 0)
                {
                   
                    Tag packet = null;
                    lock (lockObj)
                        packet = inventoryTagQueue.Dequeue();

                    if (packet != null)
                    {
                        try
                        {
                            String epc = packet.EPC;
                            if (tagList.ContainsKey(epc))
                            {
                                #region 更新列表
                                ListViewItem listRow = tagList[epc];
                                ListViewItem updateRow = null;
                                foreach (ListViewItem row in this.tagListView.Items)
                                {
                                    if (listRow == row)
                                    {
                                        updateRow = row;
                                        break;
                                    }
                                }
                                if (updateRow != null)
                                {

                                    string s = updateRow.SubItems[1].Text;
                                    UInt64 countValue = UInt64.Parse(s) + 1;
                                    updateRow.SubItems[1].Text = countValue.ToString();
                                    updateRow.SubItems[2].Text = packet.RSSI.ToString();
                                }
                                #endregion
                            }
                            else
                            {
                                #region 新增列表
                                int count = this.tagListView.Items.Count;
                                ListViewItem item = new ListViewItem(packet.EPC);
                                item.SubItems.Add("1");
                                item.SubItems.Add(packet.RSSI.ToString());

                                this.tagListView.Items.Add(item);

                                this.tagList.Add(packet.EPC, item);
                                this.actual_read_count++;
                                this.TagCountTb.Text = this.actual_read_count.ToString();
                                this.TagCountTb.Update();
                                #endregion
                            }
                        }
                        catch (Exception ex)
                        {

                        }
                    }
                }//while循环
            };

            this.Invoke(updateList);
        }

        private void stopBtn_Click(object sender, EventArgs e)
        {
            jwReader.RFID_Stop_Inventory();//停止当前UHF操作
            jwReader.TagsReported -= TagsReport;
            stopInventoryFlag = true;
        }



        /// <summary>
        /// 清空盘点
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void clearBtn_Click(object sender, EventArgs e)
        {
            clearInventoryData();
        }





        private void tagListView_DoubleClick(object sender, EventArgs e)
        {
             ListView.SelectedIndexCollection sic= this.tagListView.SelectedIndices;
            if (sic==null||sic.Count <= 0)
                ;
            else
            {
                int index=this.tagListView.SelectedIndices[0];
                String epc = this.tagListView.Items[index].SubItems[0].Text;

                //跳转到读写界面并赋值
                AccessForm af = new AccessForm();
                af.SetJWReader(jwReader);
                af.setEPC(epc);
                //隐藏当前视窗
                this.Hide();
                af.ShowDialog();
                //关闭视窗后
                this.Show();
            }
        }
    }
}